<!doctype html><html lang="en-US"><!--  --><meta http-equiv="content-type" content="text/html;charset=UTF-8" /><!-- / -->
<head><meta charset="UTF-8"><meta name="viewport" content="width=device-width, initial-scale=1">  <link media="all" href="../../wp-content/cache/autoptimize/4/css/autoptimize_dc56928889f08ceb40a4520f8ece953b.css" rel="stylesheet" /><title>C# &#8211; Should a Connect() Method return false on failure or throw exception &#8211; free-tor-game.com</title> <script>MathJax = {
  tex: {
    inlineMath: [['$','$'],['\\(','\\)']], 
    processEscapes: true
  },
  options: {
    ignoreHtmlClass: 'tex2jax_ignore|editor-rich-text'
  }
};</script> <link rel='dns-prefetch' href='http://cdn.jsdelivr.net/' /><link rel='dns-prefetch' href='http://cdnjs.cloudflare.com/' /><link rel='dns-prefetch' href='http://stackpath.bootstrapcdn.com/' /><link rel='dns-prefetch' href='http://s.w.org/' /><link rel='stylesheet' id='highlight-css'  href='../../../cdnjs.cloudflare.com/ajax/libs/highlight.js/11.2.0/styles/vs2015.min40df.css?ver=5.6' type='text/css' media='all' /><link rel='stylesheet' id='bootstrap-css'  href='../../../cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.mina352.css?ver=4.1.3' type='text/css' media='all' /><link rel='stylesheet' id='icons-css'  href='../../../stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min40df.css?ver=5.6' type='text/css' media='all' /> <script type='text/javascript' src='../../../cdnjs.cloudflare.com/ajax/libs/jquery/1.12.4/jquery.minb8ff.js?ver=1.12.4' id='jquery-js'></script> <script type='text/javascript' src='../../../cdnjs.cloudflare.com/ajax/libs/highlight.js/11.2.0/highlight.minc9a3.js?ver=11.2.0' id='highlight-js'></script>            <script>hljs.highlightAll();</script> <script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-1731162805004860" crossorigin="anonymous"></script></head><body class="software-template-default single single-software postid-1079095 no-sidebar"><div id="page" class="site"> <a class="skip-link screen-reader-text" href="#content">Skip to content</a><header id="topnav" class="sticky-top"><nav class="navbar navbar-expand-md navbar-light bg-light" role="navigation"> <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-primary" aria-controls="navbar-primary" aria-expanded="false" aria-label="Toggle navigation"> <span class="navbar-toggler-icon"></span> </button> <a class="navbar-brand" href="#">free-tor-game.com</a><div id="navbar-primary" class="collapse navbar-collapse"><ul id="menu-primary-top-menu" class="nav navbar-nav"><li itemscope="itemscope" itemtype="https://www.schema.org/SiteNavigationElement" id="menu-item-192" class="menu-item menu-item-type-taxonomy menu-item-object-category menu-item-has-children dropdown menu-item-192 nav-item"><a title="Tools" href="#" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" class="dropdown-toggle nav-link" id="menu-item-dropdown-192">Tools</a><ul class="dropdown-menu" aria-labelledby="menu-item-dropdown-192" role="menu"><li itemscope="itemscope" itemtype="https://www.schema.org/SiteNavigationElement" id="menu-item-193" class="menu-item menu-item-type-post_type menu-item-object-post menu-item-193 nav-item"><a title="ARFCN-Frequency Converter" href="../../3gpp-arfcn-frequency-converter/" class="dropdown-item">ARFCN-Frequency Converter</a></li></ul></li><li itemscope="itemscope" itemtype="https://www.schema.org/SiteNavigationElement" id="menu-item-2301" class="menu-item menu-item-type-post_type menu-item-object-page menu-item-2301 nav-item"><a title="Contact Us" href="../../contact/" class="nav-link">Contact Us</a></li></ul></div></nav></header><div id="content" class="site-content"><div id="primary" class="content-area"><main id="main" class="site-main"><div id="main-container" class="container"><div id="main-row" class="row"><div id="main-col" class="col-12"><article id="post-1079095" class="post-1079095 software type-software status-publish hentry category-software tag-c tag-exceptions"><div class="row"><div class="col-12 col-md-8"><header class="entry-header"><h1 class="entry-title text-danger">C# &#8211; Should a Connect() Method return false on failure or throw exception</h1></header><p style='font-size:1.2em'><span class="mr-2 badge badge-success">c</span><span class="mr-2 badge badge-info">exceptions</span></p><div class="entry-content"><p>I found this question<br /> <a href="https://softwareengineering.stackexchange.com/questions/159096/return-magic-value-throw-exception-or-return-false-on-failure/159105">Return magic value, throw exception or return false on failure?</a>, however it is kind of broad. I would like a more specific answer.</p><p>Say I have a method called Connect() which connects to a remote device over an Ethernet socket. If the socket fails it throws SocketException.</p><p>So the way I see it I have 2 valid options.</p><p>Option 1: Capture exception and return true.</p><pre><code>public void CallConnect()
{
    while(!Connect())
    {
        //** Log error 
    }

    //** do what ever
}


public bool Connect()
{
    try
    {
        _socket.Connect()
        return true;
    }
    catch(SocketException)
    {
        return false;
    }
    catch(&lt;SomeOtherException1&gt;)
    {
        return false;
    }
    catch(&lt;SomeOtherException2&gt;)
    {
        return false;
    }   
}
</code></pre><p>Option 2: Throw Generic Exception</p><pre><code>public void Connect()
{
    try
    {
        _socket.Connect()
    }
    catch(SocketException ex)
    {
        throw new ConnectionException(&#34;Blah blah I failed 1&#34;, ex); 
    }
    catch(&lt;SomeOtherException1&gt;)
    {
        throw new ConnectionException(&#34;Blah blah I failed 2&#34;, ex);
    }
    catch(&lt;SomeOtherException2&gt;)
    {
        throw new ConnectionException(&#34;Blah blah I failed 3&#34;, ex);
    }   
}

public void CallConnect()
{
    bool connected == false;
    while(!connected)
    {
        try
        {
            Connect();
            connected = true;
        }
        catch(ConnectionException)
        {
            //** Log error 
        }

    }
}
</code></pre><p>My understanding if a failure is a normal or expected behavior, that no exception should be thrown, but if failure should never happen then yes an exception should be thrown.</p><p>Failing to connect could be considered expected behavior if for enable someone unplugs the Ethernet cable, or someone turns off the device it needs to communicate with, but it could also be consider exceptional behavior too.</p><p>Now if I were to ask how should handle it if send fails after I have successfully connected, I think that answer is obviously throw an exception.</p></div>     <div id="comments" class="comments-area"><div class="row"><div class="col-12"><div class="mt-3 border-bottom border-success"><h4 class="text-success"><i class='fa fa-check-circle text-success mr-3'></i><span>Best Answer</span></h4></div><div class='bg-transparent mb-3'><p>The convention with .NET is to throw an exception, and if you want a method that returns a boolean create another method called TryFoo.</p><p>So, <code>void Connect()</code> would throw an exception for any sort of failure.</p><p>And <code>bool TryConnect()</code> would return false for any reason that Connect would throw an exception.</p><p>Examples of this convention are the various <code>parse</code> methods, like <code>int int.Parse(string)</code> which has another method called <code>bool int.TryParse(string, out int)</code>.</p><p>The <code>int.Parse(string)</code> method throws an exception if it cannot parse the string. The <code>int.TryParse(string, out int)</code> method will return false if it cannot parse the string.</p><p>The only difference between this example in the .NET framework, and your situation is the <code>out</code> argument, which you don&#39;t need in your case.</p><p>This allows you two use cases:</p><pre><code>foo.Connect();
foo.OperationThatRequiresAConnection();
</code></pre><p>The example above doesn&#39;t attempt to do any error handling. By throwing an exception you can guarantee that the next statement following a call to <code>Connect</code> will have an open connection. The code example above can&#39;t handle the case when a connection fails, and relies on higher level error handling in the application.</p><p>Contrast that with:</p><pre><code>if (foo.TryConnect())
{
    foo.OperationThatRequiresAConnection();
}
else
{
    MessageBox.Show(&#34;Your username or password was incorrect. Try again.&#34;);
}
</code></pre><p>In this fictional scenario, the code can recover from a failed connection by asking the user to reenter information and try the operation again.</p><h2>Library classes with no access to source code</h2><p>If you are dealing with a class library that has a <code>Connect()</code> method, and you don&#39;t have access to the source code to add a <code>TryConnect</code> method --- C# extension methods to the rescue!</p><pre><code>public static bool TryConnect(this LibraryClass connector)
{
    try
    {
        connector.Connect();

        return true;
    }
    catch (Exception)
    {
        return false;
    }
}
</code></pre><p>Then you can add your own custom application specific logging, if you wanted.</p></div></div><div class="col-4"></div></div></div></div><div class="col-12 col-md-4"> <ins class="adsbygoogle"
 style="display:block"
 data-ad-client="ca-pub-1731162805004860"
 data-ad-slot="1340983829"
 data-ad-format="auto"
 data-full-width-responsive="true"></ins>   <div class='mt-3 ml-4 border-bottom border-success'><h6><span>Related Question</span></h6></div><ul class='list-group list-group-flush'><li class="list-group-item"><a href='../c-return-magic-value-throw-exception-or-return-false-on-failure/'>C# &#8211; Return magic value, throw exception or return false on failure</a></li><li class="list-group-item"><a href='../c-what-to-return-with-async-crud-methods/'>C# &#8211; What to Return with Async CRUD methods</a></li><li class="list-group-item"><a href='../how-should-i-handle-logger-failures/'>How should I handle logger failures</a></li><li class="list-group-item"><a href='../c-when-to-create-a-custom-exception-in-c/'>C# &#8211; When to create a custom exception in C#</a></li><li class="list-group-item"><a href='../parameter-to-control-whether-to-throw-an-exception-or-return-null-good-practice/'>Parameter to control whether to throw an exception or return null &#8211; good practice</a></li></ul></div></div><footer class="entry-footer"></footer></article></div></div></div></main></div></div><footer id="colophon" class="site-footer"><div class="site-info"></div></footer></div>    <script type='text/javascript' src='../../../cdn.jsdelivr.net/npm/mathjax%403/es5/tex-chtml40df.js?ver=5.6' id='mathjax-js'></script> <script type='text/javascript' src='../../../cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/esm/popper.min43d3.js?ver=1.14.7' id='popper-js'></script> <script type='text/javascript' src='../../../cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.min5b31.js?ver=4.3.1' id='bootstrap4-js'></script> <script type='text/javascript' id='itectec-common-js-extra'>var globalObject = {"homeUrl":"https:\/\/free-tor-game.com\/"};</script> <script defer src="../../wp-content/cache/autoptimize/4/js/autoptimize_63b39f43631847816e6b0e70ea726624.js"></script></body></html>